File: server-data/resources/[esx]/esx_inventory/server/classes/inventory.lua
ESXItemDB = {}
Inventories = {}

function CreateInventory(name, owner, items)
	local self = {}

	self.name = name
	self.owner = owner
	self.items = items or {}
	self.dirty = false

	self.onChange = function()
		self.updateWeight()
		self.dirty = true

		local xPlayer = self.getXPlayer()
		TriggerClientEvent('esx_inventory:onInventoryUpdate', xPlayer.source, self.serialize())
	end

	self.getXPlayer = function()
		if self.owner == nil then
			return nil
		end
		return ESX.GetPlayerFromIdentifier(self.owner)
	end

	self.updateWeight = function()
		local xPlayer = self.getXPlayer()
		local weight = 0.0

		for _,item in pairs(self.items) do
			if item.name == 'esx_item' then
				item.weight = ESXItemDB[item.name] or 0
			end

			if item.weigth ~= nil and item.amount ~= nil then
				weight = weight + item.weight * item.amount
			end
		end

		if xPlayer ~= nil then
			for i=1, #xPlayer.inventory, 1 do
				local item = xPlayer.inventory[i]
				if item.count > 0 then
					local itemWeight = 0
					if ESXItemDB[item.name] ~= nil then
						itemWeight = ESXItemDB[item.name].weight or 0
					end

					if itemWeight ~= nil and item.count > 0 then
						 weight = weight + item.count * itemWeight / 1000.0
					end
				end
			end
		end

		self.weight = weight
	--	print("updateWeight "..tostring(weight))
	end

	self.findItem = function (item)
		item = DeserializeItem(item)

		if item == nil then
			return nil, nil
		end

		for k,v in pairs(self.items) do
			if v.equals(item) then
				return k,v
			end
		end

		return nil, nil
	end

	self.removeItem = function(item, silent)
		local xPlayer = self.getXPlayer()
		item = DeserializeItem(item)

		if item == nil then
			print("Nil item!")
			return false
		end

		if item.amount < 0 then
			print("Wrong item amount!")
			return false
		end

		-- Pocket
		if xPlayer ~= nil then
			if item.name == "esx_item" then
				if item.extra ~= nil then
					local curitem = xPlayer.getInventoryItem(item.extra.name)
					if curitem == nil then
						return false
					end
					if curitem.count ~= nil and curitem.count >= item.amount then
						xPlayer.removeInventoryItem(item.extra.name, item.amount)
					else
						return false
					end
					self.onChange()
					return true
				end
				return false

			elseif item.name == "money" then
				if ESX.Round(xPlayer.getMoney() - item.amount, 2) >= 0 then
					xPlayer.removeMoney(item.amount)
					if not silent then
						TriggerClientEvent('esx_inventory:showItemNotification', xPlayer.source, false, item.getLabel(), item.amount)
					end
					self.onChange()
					return true
				end
				return false
			elseif item.name == "black_money" then
				if ESX.Round(xPlayer.getAccount('black_money').money - item.amount, 2) >= 0 then
					xPlayer.removeAccountMoney('black_money', item.amount)
					if not silent then
						TriggerClientEvent('esx_inventory:showItemNotification', xPlayer.source, false, item.getLabel(), item.amount)
					end
					self.onChange()
					return true
				end
				return false
			end

		end

		local old_item_index, old_item = self.findItem(item)

		if old_item_index then
			if old_item.amount - item.amount > 0 then
				self.items[old_item_index].amount = old_item.amount - item.amount
				self.onChange()
			elseif old_item.amount - item.amount == 0 then
				self.items[old_item_index] = nil
				self.onChange()
			else
				return false
			end
		else
			return false
		end

		if xPlayer ~= nil and not silent then
			TriggerClientEvent('esx_inventory:showItemNotification', xPlayer.source, false, item.getLabel(), item.amount)
		end

		self.onChange()
		return true
	end

	self.addItem = function(item, silent)
		local xPlayer = self.getXPlayer()
		item = DeserializeItem(item)

		if item == nil then
			print("Nil item!")
			return false
		end

		if item.amount < 0 then
			print("Wrong item amount!")
			return false
		end

		-- Pocket
		if xPlayer ~= nil then
			if item.name == "esx_item" then
				if xPlayer ~= nil then
					if item.extra == nil then
						return false
					end

					xPlayer.addInventoryItem(item.extra.name, item.amount)
					self.onChange()
					return true
				end
				return false
			elseif item.name == "money" then
				if xPlayer then
					xPlayer.addMoney(item.amount)
					if not silent then
						TriggerClientEvent('esx_inventory:showItemNotification', xPlayer.source, true, item.getLabel(), item.amount)
					end
					self.onChange()
					return true
				end
				return false
			elseif item.name == "black_money" then
				if xPlayer  then
					xPlayer.addAccountMoney('black_money', item.amount)
					if not silent then
						TriggerClientEvent('esx_inventory:showItemNotification', xPlayer.source, true, item.getLabel(), item.amount)
					end
					self.onChange()
					return true
				end
				return false
			end
		end

		local old_item_index, old_item = self.findItem(item)

		if old_item_index then
			if item.amount + old_item.amount > 0 then
				self.items[old_item_index].amount = item.amount + old_item.amount
				self.onChange()
			elseif item.amount + old_item.amount == 0 then
				self.items[old_item_index] = nil
				self.onChange()
			else
				return false
			end
		elseif item.amount > 0 then
			table.insert(self.items, item)
			self.onChange()
		end

		if xPlayer ~= nil and not silent then
			TriggerClientEvent('esx_inventory:showItemNotification', xPlayer.source, true, item.getLabel(), item.amount)
		end

		return true
	end

	self.save = function(force)
		local xPlayer = self.getXPlayer()

		if xPlayer == false then
			return false
		end

		if self.dirty == false and not force then
			return false
		end

		local ser = self.serialize()

		TriggerEvent('esx_datastore:getDataStore', 'esx_inventory', xPlayer.identifier, function(store)
			store.set(name, ser.items)
		end)

		self.dirty = false
		return true
	end

	self.serialize = function()
		local ser = {}
		ser.name = self.name
		ser.owner = self.owner

		ser.items = {}
		for k,v in pairs(self.items) do
			table.insert(ser.items, v.serialize())
		end

		return ser
	end

	self.weight = self.updateWeight()
	return self

end

function DeserializeInventory(name, owner, items)
	local inv = CreateInventory(name, owner, {})

	for k,v in pairs(items) do
		inv.addItem(DeserializeItem(v), true)
	end

	return inv
end

function LoadInventory(name, owner)
	TriggerEvent('esx_datastore:getDataStore', 'esx_inventory', owner, function(store)
		local items = store.get(name) or {}

		getInventory(name, owner) -- create new inventory if not exists
		Inventories[name][owner] = DeserializeInventory(name, owner, items) -- replace new iunventory
	end)
end

function loadESXItemDB()
	ESXItemDB = {}
	local items = MySQL.Sync.fetchAll('SELECT * FROM items')
	for i=1,#items do
		local item = items[i]
		ESXItemDB[item.name] = item
	end
end

function getInventory(name, owner)
	if Inventories[name] == nil then
		Inventories[name] = {}
	end

	if Inventories[name][owner] == nil then
		Inventories[name][owner] = CreateInventory(name, owner, {})
	end

	return Inventories[name][owner]
end

File: server-data/resources/[esx]/esx_inventory/server/classes/inventory.js
let Inventories = [];

class Inventory {
	constructor(name, owner, items) {
		this.name = name;
		this.owner = owner;

		if (items) {
			this.items = items;
		} else {
			this.items = new Array();
		}

		this.dirty = false;
		this.weight = this.updateWeight()
	};

	onChange() {
		this.updateWeight();
		this.dirty = true;

		var xPlayer = this.getXPlayer();
		TriggerClientEvent('esx_inventory:onInventoryUpdate', xPlayer.source, this.serialize());
//		console.log("onChange: " + JSON.stringify(this.serialize()));
	}

	getXPlayer() {
		if (this.owner == null) {
			return null;
		}
		return ESX.GetPlayerFromIdentifier(this.owner);
	}

	updateWeight() {
		var xPlayer = this.getXPlayer()
		var weight = 0.0

		for (var k in this.items) {
			var item = this.items[k];
			if (item.weight != null && item.amount != null) {
				weight = weight + item.weight * item.amount
			}
		}

/*
		if (xPlayer != null) {
			for (var k in xPlayer.inventory) {
				var item = xPlayer.inventory[k]
				if (item.count > 0) {
					var itemWeight = 0
					if (global.ESXItemDB[item.name] != null) {
						itemWeight = global.ESXItemDB[item.name].weight || 0
					}

					if (itemWeight != null && item.count > 0) {
						 weight = weight + item.count * itemWeight / 1000.0
					}
				}
			}
		}
*/
		this.weight = weight
//		print(weight)
	}

	findItem(item) {
		item = DeserializeItem(item)

		if (item == null) {
			return null;
		}

		for (var k in this.items) {
			if (this.items[k].equals(item)) {
				return this.items[k];
			}
		}

		return null
	}

	findItemIndex(item) {
		item = DeserializeItem(item)

		if (item == null) {
			return null;
		}

		for (var k in this.items) {
			if (this.items[k].equals(item)) {
				return k;
			}
		}

		return null
	}

	removeItem(item, silent) {
		var xPlayer = this.getXPlayer()
		item = DeserializeItem(item)

		if (item == null) {
			console.log("null item!")
			return false
		}

		if (item.amount < 0) {
			console.log("Wrong item amount!")
			return false
		}

		var res = item.removeItem(this)
		if (res == true) {
			this.onChange()
		}

		return res
	}

	addItem(item, silent) {
		var xPlayer = this.getXPlayer();
		var item = DeserializeItem(item);

		if (item == null) {
			console.log("null item!")
			return false
		}

		if (item.amount == null || item.amount < 0) {
			console.log("Wrong item amount!")
			return false
		}

		var res = item.addItem(this)
		if (res == true) {
			this.onChange()
		}

		return res
	}

	save(force) {
//		console.log("Trying to save!")
		if (this.dirty == false && !force) {
			return false;
		}

		var ser = this.serialize(true);

		let invname = 'esx_inventory';
		if (this.name != 'pocket') {
			invname = 'esx_inventory_' + this.name
		}

		TriggerEvent('esx_datastore:getDataStore', invname, this.owner, function(store) {
			if (store !== undefined) {
				store.set(ser.name, ser.items);
			} else {
				print("Unable to get datastore with name '"+this.name+"' and owner '"+this.owner+"'")
			}
		});

		this.dirty = false;
		return true;
	}

	serialize(forSaving) {
		var ser = {}
		ser.name = this.name
		ser.owner = this.owner
		ser.weight = this.weight

		ser.items = []
		for (var i in this.items) {
			if (!forSaving || this.items[i].saveable || this.name != "pocket") {
				var item_ser_str = this.items[i].serialize()
				if (item_ser_str != null) {
					ser.items.push(item_ser_str)
				}
			}
		}

		return ser;
	}

	removeItems(name) {
		var items2 = []

		for (var k in this.items) {
            var item = this.items[k]
			if (item.name != name) {
				items2.push(item);
			}
		}

		this.items = items2;
	}

	updateLoadout() {
		this.removeItems("equipped_weapon")
		var loadout = getInventoryLoadout(this.name, this.owner);
		for (var k in loadout) {
			var item = DeserializeItem(loadout[k]);
			if (item != null) {
				this.items.push(item);
			}
		}

		return this.loadout;
	}

	updateESXItems() {
		this.removeItems("esx_item")
		var xPlayer = this.getXPlayer();
		var playerData = xPlayer

		for (var i in playerData.inventory) {
			if (playerData.inventory[i].count > 0) {
				var esx_item = playerData.inventory[i]
				var item = {}
				item.name = "esx_item"
				item.extra = {
					name:      esx_item.name,
					label:     esx_item.label,
					usable:    esx_item.usable,
					rare:      esx_item.rare,
					canRemove: esx_item.canRemove,
				}
				item.amount = playerData.inventory[i].count
				item.droppable = playerData.inventory[i].canRemove

				item = DeserializeItem(item);
				if (item != null) {
					this.items.push(item);
				}
			}
		}
	}

	updateMoneyItems() {
		var xPlayer = this.getXPlayer();

		this.removeItems("money")
		this.removeItems("account_money")
		this.removeItems("black_money")

		var item = {}
		item.name = 'money'
		item.extra = {}
		item.amount = xPlayer.getMoney()
		item.droppable = true
		item = DeserializeItem(item);
		if (item != null) {
			this.items.push(item);
		}
		
	        var account = xPlayer.getAccount("bank")
		item = {}
		item.name = 'account_money'
		item.extra = {"account_name": "bank"}

		item.amount = account.money
		item.droppable = false
		item = DeserializeItem(item);
		if (item != null) {
			this.items.push(item);
		}

	        var account = xPlayer.getAccount("black_money")
		item = {}
		item.name = 'black_money'
		item.extra = {"account_name": "black_money"}

		item.amount = account.money
		item.droppable = true
		item = DeserializeItem(item);
		if (item != null) {
			this.items.push(item);
		}
	}

}

function DeserializeInventory(name, owner, items) {
	var inv = new Inventory(name, owner, [])

	for (var k in items) {
		inv.addItem(DeserializeItem(items[k]), true);
	}

	return inv;
}

function LoadInventoryIfNotExists(name, owner) {
	if (Inventories[name] !== undefined && Inventories[name][owner] !== undefined) {
		return Inventories[name][owner];
	}

	if (Inventories[name] == null) {
		Inventories[name] = []
	}

	let invname = 'esx_inventory';
	if (name != 'pocket') {
		invname = 'esx_inventory_' + this.name
	}

	var items = [];
	var store = exports["esx_datastore"]["GetDataStore"](invname, owner, true)
	if (store) {
		items = store.get(name) || [];
	}

	Inventories[name][owner] = DeserializeInventory(name, owner, items) // replace new inventory
	Inventories[name][owner].onChange()

	return Inventories[name][owner]
}

function LoadInventory(name, owner) {
	TriggerEvent('esx_datastore:getDataStore', 'esx_inventory', owner, function(store) {
		var items = store.get(name) || [];

		getInventory(name, owner) // create new inventory if (not exists
		Inventories[name][owner] = DeserializeInventory(name, owner, items) // replace new inventory
		Inventories[name][owner].onChange()
	})
}

function getInventoryLoadout(name, owner) {
	var xPlayer = ESX.GetPlayerFromIdentifier(owner);
	if (xPlayer == null) {
		return {}
	}

	var loa = [];

	for (i in xPlayer.loadout) {
		var weapon = xPlayer.loadout[i];
		var item = {};
		item.name = "equipped_weapon";
		item.extra = {
			"weapon_name": weapon.name,
			"ammo": weapon.ammo,
			"weapon_label": weapon.label,
        };

		item.label = weapon.label;
		loa.push(item);
	}

	return loa;
}


function getInventory(name, owner) {
	if (Inventories[name] == null) {
		Inventories[name] = []
	}

	if (Inventories[name][owner] == null) {
		Inventories[name][owner] = new Inventory(name, owner, [])
	}

	var inv = Inventories[name][owner]
	inv.updateLoadout();
	inv.updateESXItems();
	inv.updateMoneyItems();
	inv.updateWeight();
	inv.items.sort(function(a,b) {
	var s1 = b.priority - a.priority
	if (s1 == 0 && b.weight && a.weight && b.amount && a.amount) {
		s1 = b.weight*b.amount - a.weight*a.amount
	}

        return s1;
    })

	return inv;
}
